src/hardware/driver/dac.rs

```diff
@@ -0,0 +1,245 @@
+        pub const SPAN_17_5V: u32 = 0b0111 << 6;
+        pub const SPAN_20V: u32 = 0b1000 << 6;
+        pub const SPAN_22_5V: u32 = 0b1001 << 6;
+        pub const SPAN_25V: u32 = 0b1010 << 6;
+        pub const SPAN_27_5V: u32 = 0b1011 << 6;
+        pub const SPAN_30V: u32 = 0b1100 << 6;
+    }
+    pub mod PDN {
+        pub const POWERUP: u32 = 0b0 << 4;
+        pub const POWERDOWN: u32 = 0b1 << 4;
+    }
+}
+
+#[allow(non_snake_case)]
+pub mod TRIGGER {
+    pub mod RCLTMP {
+        pub const RECAL: u32 = 0b1 << 8;
+    }
+    pub mod SRST {
+        pub const RESET: u32 = 0b1 << 6;
+    }
+    pub mod SCLR {
+        pub const CLEAR: u32 = 0b1 << 5;
+    }
+}
+
+#[allow(non_snake_case)]
+pub mod STATUS {
+    pub mod ALM {
+        pub const RECALIBRATED: u32 = 0b1 << 12;
+    }
+}
+
+#[allow(non_snake_case)]
+pub mod CONFIG2 {
+    pub mod DIS_TNH {
+        pub const DISABLED: u32 = 0b0 << 7;
+        pub const ENABLED: u32 = 0b1 << 7;
+    }
+    pub mod UP_RATE {
+        pub const UPDATE_RATE_1_MHZ_38_MHZ_SCLK: u32 = 0b000 << 4;
+        pub const UPDATE_RATE_0_9_MHZ_34_MHZ_SCLK: u32 = 0b001 << 4;
+        pub const UPDATE_RATE_0_8_MHZ_31_MHZ_SCLK: u32 = 0b010 << 4;
+        pub const UPDATE_RATE_1_2_MHZ_45_MHZ_SCLK: u32 = 0b011 << 4;
+        pub const UPDATE_RATE_0_5_MHZ_21_MHZ_SCLK: u32 = 0b100 << 4;
+        pub const UPDATE_RATE_0_45_MHZ_18_MHZ_SCLK: u32 = 0b101 << 4;
+        pub const UPDATE_RATE_0_4_MHZ_16_MHZ_SCLK: u32 = 0b110 << 4;
+        pub const UPDATE_RATE_0_6_MHZ_24_MHZ_SCLK: u32 = 0b111 << 4;
+    }
+}
+
+/// DAC value out of bounds error.
+#[derive(Debug)]
+pub enum Error {
+    Bounds,
+}
+
+/// A type representing a DAC sample.
+#[derive(Copy, Clone, Debug)]
+pub struct DacCode(u32);
+impl DacCode {
+    // DAC constants
+    const MAX_DAC_WORD: i32 = 1 << 20; // maximum DAC dataword (exclusive)
+    const VREF_DAC: f32 = 10.0; // Difference between positive and negaitiv reference pin
+}
+
+impl TryFrom<(f32, ChannelVariant)> for DacCode {
+    type Error = Error;
+    /// Convert an f32 representing a current int the corresponding DAC output code for the respective channel.
+    fn try_from(current_channel: (f32, ChannelVariant)) -> Result<Self, Error> {
+        let (current, channel) = current_channel;
+        if channel == ChannelVariant::HighPowerAnodeGrounded
+            || channel == ChannelVariant::LowNoiseAnodeGrounded
+        {
+            if current <= 0.0 {
+                return Err(Error::Bounds);
+            }
+        } else if current < 0.0 {
+            return Err(Error::Bounds);
+        }
+        let dac_code = (channel.transimpedance()
+            * current
+            * (DacCode::MAX_DAC_WORD as f32 / DacCode::VREF_DAC))
+            as i32;
+
+        if !(-DacCode::MAX_DAC_WORD..DacCode::MAX_DAC_WORD).contains(&dac_code)
+        {
+            return Err(Error::Bounds);
+        };
+
+        Ok(Self((dac_code & 0xfffff) as u32)) // translate to valid DAC code
+    }
+}
```

Comment:
Something like this:

```suggestion
    fn try_from(current_channel: (f32, ChannelVariant)) -> Result<Self, Error> {
        let (current, channel) = current_channel;
        let scale = channel.transimpedance()
            * (DacCode::MAX_DAC_WORD as f32 / DacCode::VREF_DAC);
        let mut code = (current * scale) as i32;
        if scale < 0. {
            // Flip the sign and overflow bits
            code ^= !(DacCode::MAX_DAC_WORD - 1);
        }
        if !(0..DacCode::MAX_DAC_WORD).contains(&code) {
            Err(Error::Bounds)
        } else {
            Ok(Self((code & (DacCode::MAX_DAC_WORD - 1)) as u32))
        }
    }
}
```

Reply:
Ah, I was missing line 150 and thought about doing it without an if.